pytorch/cppbuild.sh

```diff
@@ -181,22 +203,37 @@ TORCH_API std::ostream& operator<<(std::ostream& stream, const nn::Module& modul
 ' torch/csrc/api/include/torch/nn/module.h
 sedinplace 's/char(\(.*\))/\1/g' torch/csrc/jit/serialization/pickler.h
 
+# some windows header defines a macro named "interface"
+sedinplace 's/const std::string& interface)/const std::string\& interface_name)/g' torch/csrc/distributed/c10d/ProcessGroupGloo.hpp
+
+# Remove pytorch adaptations of FindOpenMP.cmake that.
+# On Windows without iomp and with new versions of VS 2019, including -openmp:experimental and libomp, causes
+# final binary to be linked to both libomp and vcomp and produce incorrect results.
+# Wait for eventual upstream fix, or for cmake 2.30 that allows to choose between -openmp and -openmp:experimental
+# and see if choosing experimental works. See Issue #1503.
+# On Linux, pytorch FindOpenMP.cmake picks llvm libomp over libgomp. See Issue #1504.
+# On MacOS CMake standard version works tooL
+rm cmake/Modules/FindOpenMP.cmake
+sedinplace 's/include(${CMAKE_CURRENT_LIST_DIR}\/Modules\/FindOpenMP.cmake)/find_package(OpenMP)/g' cmake/Dependencies.cmake
+
 #USE_FBGEMM=0 USE_KINETO=0 USE_GLOO=0 USE_MKLDNN=0 \
 "$PYTHON_BIN_PATH" setup.py build
 
 rm -Rf ../lib
+if [[ ! -e torch/include/gloo ]]; then
+    ln -sf ../../third_party/gloo/gloo torch/include
+fi
 ln -sf pytorch/torch/include ../include
 ln -sf pytorch/torch/lib ../lib
 ln -sf pytorch/torch/bin ../bin
 
-# fix library with correct rpath on Mac
 case $PLATFORM in
     macosx-*)
-        cp /usr/local/lib/libomp.dylib ../lib/libiomp5.dylib
-        chmod +w ../lib/libiomp5.dylib
-        install_name_tool -id @rpath/libiomp5.dylib ../lib/libiomp5.dylib
-        install_name_tool -change @rpath/libomp.dylib @rpath/libiomp5.dylib ../lib/libtorch_cpu.dylib
+        cp $(brew ls libomp|grep libomp.dylib) ../lib
```

Comment:
Please revert that. This is necessary to make it work with MKL

Reply:
There is no MKL anymore on MacOSX.

Reply:
Well, there's still an old version I suppose. Anyway, any harm in leaving this as is? Actually OpenMP is PITA, and I like to disable it everywhere I can. Does PyTorch now have it's own threading library that we can use instead of OpenMP?

Reply:
> OpenMP is PITA

Right, I spent a lot of time to make OpenMP work in all three platforms. But I think it's ok now. What is important is to prevent different OpenMP libraries to be linked to same executable, because this can silently cause incorrect computations, like what happened in #1503. MKL is supposed to detect this and throws an exception when loaded. I guess this hack making LLVM lib disguise as Intel lib is meant to prevent MKL to throw this exception. But then we cannot know which lib is actually used and I doubt MKL will work perfectly with llvm lib.

MKL adds more PITA, and I don't think we can ship a pytorch presets that can work with or without MKL.
In official binary, they chose to force MKL(link statically) on platforms where it's available, using intel omp shipped with MKL. Either we do something similar or we don't use MKL at all.

>  Does PyTorch now have it's own threading library that we can use instead of OpenMP?
No, it can use TBB instead of OpenMP, but I think TBB is more or less abandoned.
There is a native mechanism for inter-op threading in ATen  but it doesn't replace OpenMP for other threading tasks.

Reply:
So, anyway this hack does work fine with MKL, so please revert this

Reply:
Why no doing something similar for other platforms then ?

Reply:
GCC and VS don't use the same library so they don't conflict, or didn't used to anyway. That's the point here, please don't try fix what isn't broken. These lines aren't broken. Don't fix them

Reply:
On Linux, a user linking with MKL could set `MKL_THREADING_LAYER=GNU`. 
On Windows, I don't think MKL can use `vcomp` which is the OpenMP we currently linked with. So I don't know what happens if MKL is loaded. It will probably not use any threading layer at all.

I restored the hack. It was broken because the `install_name_tool -change` command supposed the installed path to `libomp` is `@rpath/libomp.dylib`, which is the case only if we run in the github runner where we changed the rpath. I fixed that. Also it didn't work on recent versions of MacOS because signature was incorrect after `install_name_tool`. I fixed that too.